pkg/controller/dns/external_dns_crd.go

```diff
@@ -0,0 +1,118 @@
+	return ExternalDNSCRDControllerName.AddToController(ctrl.NewControllerManagedBy(manager).
+		For(&v1alpha1.ExternalDNS{}).
+		Owns(&appsv1.Deployment{}).
+		Owns(&corev1.ConfigMap{}), manager.GetLogger()).
+		Complete(&ExternalDNSCRDController{
+			config: &config,
+			client: manager.GetClient(),
+			events: manager.GetEventRecorderFor("aks-app-routing-operator"),
+		})
+}
+
+func (e *ExternalDNSCRDController) Reconcile(ctx context.Context, req ctrl.Request) (res ctrl.Result, err error) {
+	defer func() {
+		metrics.HandleControllerReconcileMetrics(ExternalDNSCRDControllerName, res, err)
+	}()
+
+	// set up logger
+	logger, err := logr.FromContext(ctx)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("creating logger: %w", err)
+	}
+	logger = ExternalDNSCRDControllerName.AddToLogger(logger).WithValues("namespace", req.Namespace, "name", req.Name)
+
+	obj := &v1alpha1.ExternalDNS{}
+	if err = e.client.Get(ctx, req.NamespacedName, obj); err != nil {
+		logger.Error(client.IgnoreNotFound(err), "failed to get externaldns object, will ignore not found error")
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// verify serviceaccount
+	if _, err = keyvault.GetServiceAccountAndVerifyWorkloadIdentity(ctx, e.client, obj.GetInputServiceAccount(), obj.GetNamespace()); err != nil {
+
+		var userErr keyvault.UserError
+		if errors.As(err, &userErr) {
+			logger.Info("failed to verify service account due to user error, sending warning event: " + userErr.UserError())
+			e.events.Eventf(obj, corev1.EventTypeWarning, "FailedUpdateOrCreateExternalDNSResources", "failed serviceaccount verification: %s", userErr.UserError())
+			return ctrl.Result{}, nil
+		}
+		logger.Error(err, "failed to verify service account")
+		return ctrl.Result{}, err
+	}
+
+	inputDNSConf := buildInputDNSConfig(obj)
+	manifestsConf, err := manifests.NewExternalDNSConfig(e.config, inputDNSConf)
+	if err != nil {
+		logger.Error(err, "failed to generate ExternalDNS resources from ExternalDNS CR")
+		e.events.Eventf(obj, corev1.EventTypeWarning, "FailedUpdateOrCreateExternalDNSResources", "failed to generate external DNS resources: %s", err)
+		return ctrl.Result{}, err
+	}
+
+	resources := manifestsConf.Resources()
+
+	multiError := &multierror.Error{ErrorFormat: func(errs []error) string {
+		errStrings := make([]string, 0, len(errs))
+		for err := range errs {
+			errStrings = append(errStrings, errs[err].Error())
+		}
+		return strings.Join(errStrings, ", ")
+	}}
+
+	for _, resource := range resources {
+		resource.SetOwnerReferences([]metav1.OwnerReference{{
+			APIVersion: obj.APIVersion,
+			Controller: util.ToPtr(true),
+			Kind:       obj.Kind,
+			Name:       obj.Name,
+			UID:        obj.UID,
+		}})
+
+		currentResourceErr := util.Upsert(ctx, e.client, resource)
+		if currentResourceErr != nil {
+			logger.Error(currentResourceErr, "failed to upsert externaldns resources")
+		}
+		multiError = multierror.Append(multiError, currentResourceErr)
```

Comment:
i think this is an interesting pattern where we don't exit early and keep upserting resources if there's a failure. I think this is good but this is different from our other controllers

Reply:
My kneejerk opinion is that just exiting early is better because it's simpler but trying to think of the pros and cons. It's hard to think of a scenario where upserting all even though there's an error is helpful

Reply:
I think it would help diagnose the issue more quickly. If there's a partial failure, it's easier to pinpoint which resource failed to deploy, and the error then contains all failed resources instead of just the first one that hit failure